cloudpickle/cloudpickle.py

```diff
@@ -471,577 +474,61 @@ def _is_parametrized_type_hint(obj):
 
     def _create_parametrized_type_hint(origin, args):
         return origin[args]
+else:
+    _is_parametrized_type_hint = None
+    _create_parametrized_type_hint = None
+
+
+def parametrized_type_hint_getinitargs(obj):
+    # The distorted type check sematic for typing construct becomes:
+    # ``type(obj) is type(TypeHint)``, which means "obj is a
+    # parametrized TypeHint"
+    if type(obj) is type(Literal):  # pragma: no branch
+        initargs = (Literal, obj.__values__)
+    elif type(obj) is type(Final):  # pragma: no branch
+        initargs = (Final, obj.__type__)
+    elif type(obj) is type(ClassVar):
+        initargs = (ClassVar, obj.__type__)
+    elif type(obj) is type(Generic):
+        parameters = obj.__parameters__
+        if len(obj.__parameters__) > 0:
+            # in early Python 3.5, __parameters__ was sometimes
+            # preferred to __args__
+            initargs = (obj.__origin__, parameters)
 
-
-class CloudPickler(Pickler):
```

Comment:
Also: should we try to preserve backward-compatibility for semi-public `cloudpickle` objects? For instance, code like
`from cloudpickle.cloudpickle import CloudPickler` fails in this PR.

Reply:
I would not mind, as long as those symbols are not used in reducers (which is now tested by the old_pickles thingy).

Reply:
Actually the following used to work and is causing Ray to break on this PR:

```python
>>> from cloudpickle.cloudpickle import dumps                                                          
Traceback (most recent call last):
  File "<ipython-input-3-9394653b7d6c>", line 1, in <module>
    from cloudpickle.cloudpickle import dumps
ImportError: cannot import name 'dumps' from 'cloudpickle.cloudpickle' (/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py)
```